### PR TITLE
Feature/impersonation with support pin

### PR DESCRIPTION
**Pull request by sunilsabatp**
Feature ->1.added new field for impersonation form (i.e support pin).
                 2.allow user to change their support pin.
Refactor ->1.impersonation feature code
             

exception -> For entering wrong pin it'll not throw error. since , it throws error for such scenario (**wrong pin**)only on production . if you want to check this scenario then use support  pin `1234 `

### CODE REVIEW

**Comment by mariahosfeld** [comment]
Functionality is looking good so far.

**Comment by mariahosfeld** [comment]
Issue: The link to select the language is also present for "support pin", it needs to be removed there.

**Comment by mariahosfeld** [comment]
Functionality looks good to me.

File: src/utils/apiRequests/api.ts [hunk 1]
```diff
@@ -70,11 +72,11 @@ const handleApiError = (
 //  API call to private /profile endpoint
 export async function getAccountInfo(
   token: any,
-  impersonatedEmail?: string
+  impersonatedData?:ImpersonationData
```
Comment (mohitb35) [inline src/utils/apiRequests/api.ts:75]: Be consistent about naming things. `impersonationData` or `impersonatedData`?

File: src/utils/apiRequests/api.ts [hunk 2]
```diff
@@ -249,25 +277,33 @@ export async function postAuthenticatedRequest<T>(
 export async function deleteAuthenticatedRequest(
   url: any,
   token: any,
-  impersonatedEmail?: string,
   errorHandler?: Function
 ): Promise<any> {
+  const targetData = localStorage.getItem("impersonationData") ;
+  const  impersonatedDataFromLocal = JSON.parse(targetData);
```
Comment (mohitb35) [inline src/utils/apiRequests/api.ts:283]: You can probably combine these 2 lines. Don't forget to check for TS warnings - JSON.parse expects a string.

File: src/utils/apiRequests/api.ts [hunk 3]
```diff
@@ -312,25 +348,33 @@ export async function putAuthenticatedRequest<T>(
...
+        : 'en'
+    }`,
+  }
+  if(impersonatedDataFromLocal?.targetEmail ){
+    header["X-SWITCH-USER"] =  impersonatedDataFromLocal?.targetEmail 
+  }
+
+  if( impersonatedDataFromLocal?.supportPin){
+    header["X-USER-SUPPORT-PIN"] =  impersonatedDataFromLocal?.supportPin 
+  }
```
Comment (mohitb35) [inline src/utils/apiRequests/api.ts:372]: You could abstract this logic to a common function as it used in all authenticated requests.

File: src/features/common/Layout/UserLayout/UserLayout.tsx [hunk 4]
```diff
@@ -492,7 +527,7 @@ function UserLayout(props: any): ReactElement {
             className={styles.navlink}
             //logout user
             onClick={() => {
-              localStorage.removeItem('impersonatedEmail');
+              localStorage.removeItem('impersonatedData');
```
Comment (mohitb35) [inline src/features/common/Layout/UserLayout/UserLayout.tsx:530]: Be consistent with naming. Here you should remove `impersonationData` not `impersonatedData`

**Review by mariahosfeld — COMMENTED** [review]
Functionality works for me.

**Review by mohitb35 — CHANGES_REQUESTED** [review]
@sunilsabatp Please review my feedback.

File: src/features/common/Layout/UserLayout/UserLayout.tsx [hunk 5]
```diff
@@ -52,22 +59,50 @@ function LanguageSwitcher() {
...
+  //to  get new pin
+  const handleNewPin = async () => {
+    try {
+      const response = await putAuthenticatedRequest<SupportPin>(
+        '/app/profile/supportPin',
+        undefined,
+        token
+      );
+      if (response) {
+        setSupportPin(response?.supportPin);
```
Comment (mohitb35) [inline src/features/common/Layout/UserLayout/UserLayout.tsx:71]: When this is updated, you should maintain a sync with the user object in UserPropsContext.

**Comment by mariahosfeld** [comment]
Please make sure that

- [ ] The color of the refresh icon changes to green as well
- [ ] The user is informed about the purpose of the refresh icon by adding a title "get new pin"

File: src/features/common/Layout/UserLayout/UserLayout.tsx (hunk 4 above)
Comment (mohitb35) [reply src/features/common/Layout/UserLayout/UserLayout.tsx:503]: This is missed out.

**Review by mohitb35 — CHANGES_REQUESTED** [review]

**Review by mohitb35 — APPROVED** [review]

**Pull request merged**